### DISCUSSION TITLE
[Question] What’s the recommended way to access randomized external forces and torques applied to assets for manager-based workflow?

### DISCUSSION THREAD

**Discussion by diracdelta7** [isaac-sim/IsaacLab#1360] [Q&A]
**Question**  
First, thank you for the fantastic work!

I'm currently working on legged locomotion using a teacher-student architecture. In my setup, I define `external_forces_and_torques` in `velocity_env_cfg.py` as follows:

```python
# In velocity_env_cfg.py
external_forces_and_torques = ObsTerm(
    func=mdp.external_forces_and_torques,
    params={"sensor_cfg": SceneEntityCfg("robot", body_names=".*base")},
)
```

And in `mdp.observation.py`, I define the function:

```python
# In mdp.observation.py
def external_forces_and_torques(env: ManagerBasedRLEnv, asset_cfg: SceneEntityCfg) -> torch.Tensor:
    """External forces and torques applied to the body."""
    asset: Articulation = env.scene[asset_cfg.name]

    external_force_b = asset._external_force_b[:, asset_cfg.body_ids, :].squeeze(1)
    external_torque_b = asset._external_torque_b[:, asset_cfg.body_ids, :].squeeze(1)
    return torch.cat((external_force_b, external_torque_b), dim=1)
```

The issue is that `asset._external_force_b` and `asset._external_torque_b` are protected members of the `Articulation` class, and accessing them triggers a warning. Is there a more appropriate way to access these variables?

Additionally, in a teacher-student setup, it’s common to design events that randomize the environment or apply specific changes, such as forces, torques, or friction adjustments. For events like these, we may want to capture information about these randomizations as privileged observations for the teacher. However, accessing this information directly isn’t always feasible (e.g., `asset._external_force_b`), and in some cases, it can be time-consuming to compute (e.g., retrieving static friction via `asset.root_physx_view.get_material_properties()` requires CPU resources). Is there a recommended approach in IsaacLab for handling this efficiently?

**Comment by katari16** [comment]
@diracdelta7 Did you find a way of accessing the external forces as observations? I am also having the same issue currently
